frame/contracts/src/wasm/mod.rs

```diff
@@ -122,8 +128,25 @@ where
 		original_code: Vec<u8>,
 		schedule: &Schedule<T>,
 		owner: AccountIdOf<T>,
-	) -> Result<Self, &'static str> {
-		prepare::prepare_contract(original_code, schedule, owner)
+	) -> Result<Self, (DispatchError, &'static str)> {
+		let module =
+			prepare::prepare_contract(original_code, schedule, owner).map_err(|(err, msg)| {
```

Comment:
For extra type safety you could convert to `CodeVec` up front and then `prepare_contract` to accept `original_code: CodeVec`.

Reply:
IMHO the saftest is to do it as late as possible: When it is written to storage. This way we only need to do it at one place. If we moved the check up it needs to be duplicated in multiple places.

Reply:
I think it would still only be in one place, at the top level.

Reply:
You are right. I moved it.